Addition of identity-file argument to init-ssh!

markdingram opened the pull request:
This allows a specific key to be nominated for the Git SSH connection.

An example of where this can be useful:
Consider a setup where keys A & B both authenticate to GitHub but with
different access to private repositories repo-A & repo-B.

Currently the first key that successfully authenticates to Github is used for the remainder of the Git
operation. So without a means of selecting the key to use one of the
private repos will be unavailable (git error: "Repository: Not Found.")

NOTE: JSCH doesn't use the ~/.ssh/config file where this would commonly be set.

markdingram commented:
"./go test" works fine locally. Not sure why Travis isn't..

flosell commented:
Looks like you ran into a flaky test there. Haven't seen this before but I'll look into it. 

To your PR: 
Thanks, this looks like a valuable addition! I'd prefer if JGit or JSch (both might be at at fault here) just followed the ssh-config but your fix is definately a step in the right direction.

Two small things:
- The docstring says _"The fully expanded file name should be provided (i.e. no ~/..)"_. We do `fs/expand-home` for the known-hosts files, I think it would be more consistent if the identity files worked the same way
- Please add a bulletpoint in the [CHANGELOG.md](https://github.com/flosell/lambdacd-git/blob/master/CHANGELOG.md) so that people upgrading are aware of changes and new features

Other than that, this looks ready to merge.

markdingram commented:
I now see that both JSCh and JGit do have their own SSH config parsing... I'll do some digging to see why that isn't being used in my case. Would be preferable to fix that instead of this patch.

flosell commented:
@markdingram Any progress on your investigation? I don't want you to feel like I'm blocking this PR so if there's no short term fix in JSch or JGit, let's move forward on your workaround and worry about the rest later.

markdingram commented:
Hi @flosell - I took another look through the  code upstream in JSch and JGit to see if this would be possible another way, and for me at least there is no obvious way.

I think the upstream repo to target would be https://github.com/ymnk/jsch-agent-proxy , I may pursue a change there to allow filtering its RemoteIdentityRepository for a specific identity.

In the meantime if you are still keen I will make the changes to the PR suggested above..

flosell commented:
I agree. I'd still be happy to merge the PR

markdingram committed:
Addition of identity-file argument to init-ssh!

This allows a specific key to be nominated for the Git SSH connection.

An example of where this can be useful:
Consider a setup where keys A & B both authenticate to GitHub but with
different access to private repositories repo-A & repo-B.

Currently the first key that successfully authenticates to Github is used for the remainder of the Git
operation. So without a means of selecting the key to use one of the
private repos will be unavailable (git error: "Repository: Not Found.")

NOTE: JSCH doesn't use the ~/.ssh/config file where this would commonly be set.

markdingram commented:
Hi @flosell - I updated the PR as requested.

I have also made a request in upstream JGit - https://git.eclipse.org/r/#/c/75607/ , if this gets accepted there then it will be possible to improve the code here - by changing to override configureJsch rather than createDefaultJSch. Unfortunately the "default" JSch instance (to which we apply overrides) isn't always used - see method JschConfigSessionFactory.getJSch.  

Thanks!

flosell commented:
Awesome, thanks for the contribution and for all the effort you are putting in!
